File: python/51_构建乘积数组.py
#!/bin/bash/ python3
#-*- coding:utf-8 -*-

import time

class Solution:
    def multiply1(self, A):
        """
        使用除法的方式，复杂度也是O(2n)
        然而题目要求不能使用除法
        """
        if A == []:
            return []
        
        tmp = 1
        for i in A:
            tmp *= i
        
        result = []
        for i in A:
            result.append(tmp // i)
        
        return result

    def multiply2(self, A):
        """
        其它思路，两层遍历，每次使用1修改当前元素，复杂度O(n^2)
        这种方法低效
        """
        if A == []:
            return []
        
        result = []
        for i in range(len(A)):
            curr = A[i]
            A[i] = 1
            tmp = 1
            for j in A:
                tmp *= j
            A[i] = curr
            result.append(tmp)
        
        return result
    
    def multiply3(self, A):
        """
        常规思路，左右延伸计算，复杂度O(n*(n-1))
        """
        if A == []:
            return []
        
        result = []
        for i in range(len(A)):
            tmp = 1
            for j in range(len(A)):
                if j != i:
                    tmp *= A[j]
            result.append(tmp)
        
        return result

    def multiply(self, A):
        """
        高效思路，两次遍历，复杂度O(2n)
        """
        if A == []:
            return []
        
        tmp = 1
        result = []
        for i in A:
            result.append(tmp)
            tmp *= i
        
        tmp = 1
        for i in range(len(A) - 1, -1, -1):
            result[i] *= tmp
            tmp *= A[i]
        
        return result

    def test(self, A):
        """
        测试函数
        """
        func_vec = [self.multiply1,
                    self.multiply2,
                    self.multiply3,
                    self.multiply]
        for func in func_vec:
            start = time.time()
            result = func(A)
            end = time.time()
            print("result: {}, time(us): {:>5.2f}".format(result, (end - start)*10**6))


def main():
    vec = [1, 2, 3, 4, 5]
    vec2 = [8, 3, 5, 2, 2, 4, 1, 6, 7]

    s = Solution()
    s.test(vec)
    print("=====")
    s.test(vec2)


if __name__ == "__main__":
    main()

File: cpp/51_构建乘积数组.cpp
#include <iostream>
#include <vector>
#include <stdio.h>
#include <sys/time.h>

class Solution {
public:
    // 使用除法的思路，复杂度也是O(2n)
    // 然而题目要求不能使用除法
    std::vector<int> multiply1(const std::vector<int>& A) {
        if (A.empty()) {
            return A;
        }

        long tmp = 1;
        for (auto i : A) {
            tmp *= i;
        }

        std::vector<int> result;
        for (auto i : A) {
            result.push_back(tmp / i);
        }
        return result;
    }

    // 其它思路，两层遍历，每次使用1修改元素，复杂度O(n^2)
    // 注意到A是const，所以只能额外使用空间（这种方法低效）
    std::vector<int> multiply2(const std::vector<int>& A) {
        if (A.empty()) {
            return A;
        }

        std::vector<int> result;
        std::vector<int> data(A.begin(), A.end());

        for (int i = 0 ; i < data.size(); i++) {
            long tmp = 1;
            int curr = data[i];
            data[i] = 1;
            for (auto j : data) {
                tmp *= j;
            }
            data[i] = curr;
            result.push_back(tmp);
        }

        return result;
    }

    // 常规思路，左右延伸计算，复杂度O(n*(n-1))
    // 第一次的解法
    std::vector<int> multiply3(const std::vector<int>& A) {
        std::vector<int> ret;
        if (A.empty()) {
            return ret;
        }
        
        for (int i = 0; i < A.size(); i++) {
            int tmp = 1;
            for (int j = 0; j < A.size(); j++) {
                if (j != i) {
                    tmp *= A[j]; 
                }
            }
            ret.push_back(tmp);
        }
        
        return ret;
    }

    // 高效思路，两次遍历，区分左右侧元素构成的乘积数组
    std::vector<int> multiply(const std::vector<int>& A) {
        if (A.empty()) {
            return A;
        }
        
        int tmp = 1;
        std::vector<int> ret;
        for (auto i : A) {
            ret.push_back(tmp);
            tmp *= i;
        }
        
        tmp = 1;
        for (int i = A.size() - 1; i >= 0; i--) {
            ret[i] *= tmp;
            tmp *= A[i];
        }
        
        return ret;
    }

    // 打印一维向量
    void print_1d_vec(std::vector<int>& vec) {
        if (vec.empty()) {
            return;
        }

        for (auto i : vec) {
            std::cout << i << " ";
        }
    }

    // 测试函数
    void test(std::vector<int>& vec) {
        std::vector<int> result;
        struct timeval start, end;
        for (auto func : this->func_vec_) {
            gettimeofday(&start, 0);
            result = (this->*func)(vec);
            gettimeofday(&end, 0);
            printf("result: ");
            this->print_1d_vec(result);
            printf(", time(us): %ld\n", end.tv_usec - start.tv_usec);
        }
    }

private:
    typedef std::vector<int> (Solution::*func_ptr)(const std::vector<int>&);
    std::vector<func_ptr> func_vec_ = {&Solution::multiply1,
                                       &Solution::multiply2,
                                       &Solution::multiply3,
                                       &Solution::multiply};
};

int main(int argc, char* argv[])
{
    std::vector<int> vec = {1, 2, 3, 4 ,5};
    std::vector<int> vec2 = {8, 3, 5, 2, 2, 4, 1, 6, 7};

    Solution s;
    s.test(vec);
    std::cout << "=====" << std::endl;
    s.test(vec2);

    return 0;
}